README.md
# deprag_downloader
## Application
This ROS node allows to automatically download curves saved by a deprag screwdriver.
To do so simply call the provided ROS service with the corresponding number.

## Dependencies
This Package was developed using ROS Melodic other versions might work but are not tested.
Selenium for Python with a Firefox driver installed.

## Setup
After installing all dependencies change the "downloadDirectory" variable in deprag_downloader_node.py to 
the file path where the csv files should be saved.
Now change the folder name of the folder you cloned from git to "deprag_downloader"
Furthermore, change the "depragIP" variable to the address matching your deprag screwdriver.

## Usage
After making sure roscore is running start the deprag_downloader_node using
```
$ rosrun deprag_downloader deprag_downloader_node.py 
```
Now a curve can be saved by calling the "/deprag_download" service.
```
$ rosservice call /deprag_download arg
```

The number with which the service is called determines what curve is downloaded.
Notice that 0 will download all files available, therefore passing 1 will download the latest curve.

When calling with a value < 0 it will download the last curve and the corresponding programm parameters and publish both under /deprag/curves.
```
$ rostopic echo /deprag/curves
```
A json file containing the program parameters and the curve values is saved in the download directory the files name is determined by the current system time.

## json
The json contains all program information in accordance with the used deprag manual. The csv curve data is saved in a 
structure called "csv" where it is separated into different arrays.

"Zeit(0.001 ms)" represents the time passed since starting the programm in 0.001 ms

"Drehmoment Messsystem 1 (Nm)" is the torque messured by system 1 in Nm

"Winkel Messsystem 1" is the angle messured by system 1 

"Drehmoment Motor (Nm)" is the torque messured by the motor in Nm

"Winkel Motor" is the angle messured by the motor

"Drehzahl (U/min)" represents rpm

"Schritt" represents the steps taken

"Stromstaerke (A)" represents the amperage drawn in amps

"Temperatur (C)" represents the motor temperature in Celcius


## Acknowledgement
Sponsored by the Ministry of the Environment Baden-Württemberg, in the context of the Strategic Dialogue Automotive Industry, and supervised by the Project Management Agency Karlsruhe (PTKA). Funding number: L7520101

https://www.ipa.fraunhofer.de/de/referenzprojekte/DeMoBat.html

## Contact
For more information please feel free to contact: <br />
M.Sc. Anwar Al Assadi<br />
Wissenschaftlicher Mitarbeiter<br />
Fraunhofer‐Institut für<br />
Produktionstechnik und Automatisierung IPA<br />
Abteilung Roboter- und Assistenzsysteme<br />
Nobelstraße 12 │ 70569 Stuttgart <br />
Telefon +49 711 970-1264 <br />
anwar.alassadi@ipa.fraunhofer.de<br />


CHANGELOG.md
## Upcoming 1.1.0<a name="1.1.0"></a> (2023-02-09)

### Docs

- add changelog template ([aa69d65](https://github.com/balandbal/deprag_screw_results/commit/aa69d65fda211b4b7a64929b33ff6e3e932abd81))

### Features

- refactor rospy node ([0897a8c](https://github.com/balandbal/deprag_screw_results/commit/0897a8ca91adfb5828ff71c6888ecadf250ab2c7))

### Style

- clean up in CMakeLists.txt ([c6f8b60](https://github.com/balandbal/deprag_screw_results/commit/c6f8b60408fe5ed62cb9fbad009864361236654b))


src/script_download_curve.cpp
#include "cppitasc/coordination/script_rosservice.hpp"

#include <deprag_downloader/download_request.h>


class ScriptRosServiceDownloadCurveRequest : public ScriptRosService<deprag_downloader::download_request, deprag_downloader::download_request::Request, deprag_downloader::download_request::Response>
{
public:
    ScriptRosServiceDownloadCurveRequest(const string& name)
        : ScriptRosService<deprag_downloader::download_request, deprag_downloader::download_request::Request, deprag_downloader::download_request::Response>(
            name)
    {
    }

    bool init(Dict& params) override {
        extract(params["retries"], retries_);
        extract(params["download_target"], download_target_);

        service_.request.iTarget = download_target_;
        return ScriptRosService::init(params);
    }

    bool call() override {
        pi_info_named("ScriptRosServiceDownloadCurveRequest::call", "Calling service: '{}'", serviceName_);
        bool result = false;

        for (int tries=0; tries <= retries_; tries++) {
            result = client_.call(service_.request, service_.response);

            pi_info_named("ScriptRosServiceDownloadCurveRequest::call", "Service response: '{}'", service_.response);
            pi_info_named("ScriptRosServiceDownloadCurveRequest::call", "Service result: '{}'", result);
            
            if (result) {
                break;
            }
        };

        return result;
    }

private:
    int download_target_{1};
    int retries_{0};
};

RUNTIME_COMPONENT(ScriptRosServiceDownloadCurveRequest)

src/deprag_downloader/downloader.py
from pathlib import Path
from time import time

import rospy
import requests as requester

from deprag_downloader.download_directory import (
    get_download_directory as _get_download_directory,
)
from deprag_downloader.browser_interaction import (
    download_curves as _download_curves,
    download_fvalues as _download_fvalues,
)
from deprag_downloader.msg import screwing
from deprag_downloader.srv import download_requestRequest, download_requestResponse


def _find_nth(haystack, needle, n):
    start = haystack.find(needle)
    while start >= 0 and n > 1:
        start = haystack.find(needle, start + len(needle))
        n -= 1
    return start


class DepragDownloader:
    def __init__(
        self, download_directory: Path, deprag_ip: str, curve_publisher: rospy.Publisher
    ) -> None:
        self.download_directory = download_directory
        self.deprag_ip = deprag_ip
        self.publisher = curve_publisher

    def __call__(self, data: download_requestRequest) -> download_requestResponse:
        pub = data.iTarget < 0
        data.iTarget = 1 if pub else data.iTarget
        download_directory = _get_download_directory(
            None if pub else self.download_directory
        )

        _download_curves(download_directory, self.deprag_ip, data.iTarget)

        if pub:
            self._publish(download_directory)

    def _publish(self, download_directory: Path) -> None:

        curve_file = next(download_directory.iterdir())

        rospy.logdebug(f'Publishing curve "{curve_file.stem}"...')

        sub_dir = (
            download_directory / "sub"
        )  # sub directory to determine what program was used
        sub_dir.mkdir(exist_ok=True)

        _download_fvalues(sub_dir, self.deprag_ip)

        with open(sub_dir / "actual.csv", "r") as info:
            _info = info.read()
        _info = _info[20:]
        c = _info.find(",")
        _info = _info[:c]

        _prgrmNumber = int(_info)

        rospy.logdebug(f'Focusing on program no. "{_prgrmNumber}"...')

        # get programm info
        _str = (
            "http://"
            + self.deprag_ip
            + "/cgi-bin/cgiread?site=getprg&program="
            + str(_prgrmNumber)
        )
        r = requester.get(_str)
        prgmInfo = r.text

        rospy.logdebug("Got program info: " + prgmInfo)

        msg = screwing()
        msg.prgrmInfo = prgmInfo

        curve_file = curve_file.rename(download_directory / "data.txt")  # type: ignore[func-returns-value]

        with open(curve_file, "r") as _csv:
            _csv_str = _csv.read()
        c = _csv_str.find("Temperatur") + 15
        _csv_str = _csv_str[c:]
        msg.csv = _csv_str

        self.publisher.publish(msg)

        filepath = self.download_directory / f"{time()}.json"

        with open(filepath, "w") as _f:
            jsonInfo = prgmInfo.replace("\n", ',\n"')
            jsonInfo = jsonInfo.replace("=", '":')
            jsonInfo = '"' + jsonInfo
            c = _find_nth(jsonInfo, ":", 2) + 1
            jsonInfo = jsonInfo[:c] + '"' + jsonInfo[c:]
            c = _find_nth(jsonInfo, ",", 2)
            jsonInfo = jsonInfo[:c] + '"' + jsonInfo[c:]
            _f.write("{ \n")
            _f.write(jsonInfo)

            zeit = "["
            Dmess1 = "["
            Wmess1 = "["
            Dmotor = "["
            Wmotor = "["
            DZ = "["
            Schritt = "["
            Strom = "["
            Temp = "["
            lines = _csv_str.count("\n") + 1  # anzahl Zeilen
            for i in range(0, lines):
                # Zeit
                c = _csv_str.find(",")
                zeit = zeit + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # Mess1 Dreh
                c = _csv_str.find(",")
                Dmess1 = Dmess1 + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # W mess1
                c = _csv_str.find(",")
                Wmess1 = Wmess1 + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # D motor
                c = _csv_str.find(",")
                Dmotor = Dmotor + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # W Motor
                c = _csv_str.find(",")
                Wmotor = Wmotor + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # Drehzahl
                c = _csv_str.find(",")
                DZ = DZ + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # Schritt
                c = _csv_str.find(",")
                Schritt = Schritt + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # Strom
                c = _csv_str.find(",")
                Strom = Strom + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # temp
                c = _csv_str.find("\n")
                Temp = Temp + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]

            zeit = zeit.replace("\n", "")

            zeit = zeit[: (len(zeit) - 2)] + "],"
            Dmess1 = Dmess1[: (len(zeit) - 2)] + "],"
            Wmess1 = Wmess1[: (len(zeit) - 2)] + "],"
            Dmotor = Dmotor[: (len(zeit) - 2)] + "],"
            Wmotor = Wmotor[: (len(zeit) - 2)] + "],"
            DZ = DZ[: (len(zeit) - 2)] + "],"
            Schritt = Schritt[: (len(zeit) - 2)] + "],"
            Strom = Strom[: (len(zeit) - 2)] + "],"
            Temp = Temp[: (len(zeit) - 2)] + "]"

            _f.write('csv" : { \n')
            t = '   "Zeit(0.001 ms)":' + zeit
            t.replace("\n", "")
            _f.write(t + "\n")
            _f.write('  "Drehmoment Messsystem 1 (Nm)":' + Dmess1 + "\n")
            _f.write('  "Winkel Messsystem 1":' + Wmess1 + "\n")
            _f.write('  "Drehmoment Motor (Nm)":' + Dmotor + "\n")
            _f.write('  "Winkel Motor":' + Wmotor + "\n")
            _f.write('  "Drehzahl (U/min)":' + DZ + "\n")
            _f.write('  "Schritt":' + Schritt + "\n")
            _f.write('  "Stromstaerke (A)":' + Strom + "\n")
            _f.write('  "Temperatur (C)":' + Temp + "\n")

            _f.write("}\n")
            _f.write("  }")

        rospy.logdebug(f'Removing "{str(download_directory)}"...')

        download_directory.unlink()


src/deprag_downloader_node.py
#!/usr/bin/env python
import os
import shutil
import time

import requests as requester
import rospy
from selenium.webdriver import Firefox
from selenium.webdriver.common.by import By
from selenium.webdriver.firefox.options import Options

from deprag_downloader.msg import screwing
from deprag_downloader.srv import download_request

downloadDirectory = "/home/adm-awi/Downloads/TestFolder"

depragIP = "172.10.25.100"  # ip address of the deprag screwdriver


def find_nth(haystack, needle, n):
    start = haystack.find(needle)
    while start >= 0 and n > 1:
        start = haystack.find(needle, start + len(needle))
        n -= 1
    return start


def _get_browser_options(download_directory: str) -> Options:
    opts = Options()

    opts.add_argument("--headless")

    opts.set_preference("browser.download.folderList", 2)
    opts.set_preference("browser.download.manager.showWhenStarting", False)
    opts.set_preference("browser.download.dir", download_directory)
    opts.set_preference("browser.helperApps.neverAsk.saveToDisk", "text/csv/ast")

    return opts


def callback_download(data):  # if data is -1 publishe curve to topic
    # copy past download part with data.iTarget as Index
    # set up headles browse
    global downloadDirectory
    backupDownload = downloadDirectory
    pub = False
    if data.iTarget < 0:
        pub = True
        data.iTarget = 1
        downloadDirectory = "/tmp/deprag_downloaderERmabs4k23lknad33"
        os.mkdir(downloadDirectory)

    opts = _get_browser_options(download_directory=downloadDirectory)

    with Firefox(options=opts) as browser:
        # interact with duck duck
        browser.get(
            "http://"
            + depragIP
            + "/cgi-bin/cgiread?site=-&request=curves&args=&mode=-1-"
        )

        dropdown = browser.find_element(By.CLASS_NAME, "dd-selected")
        dropdown.click()
        choice = browser.find_elements(By.CLASS_NAME, "dd-option")
        choice[1].click()
        downloads = browser.find_elements(By.CLASS_NAME, "download")
        downloads[data.iTarget].click()

    if pub:
        _kurve_name = os.listdir(downloadDirectory)[0]
        print(_kurve_name)
        subDir = "/sub"  # sub directory to determine what program was used
        os.mkdir(downloadDirectory + subDir)

        opts = _get_browser_options(download_directory=downloadDirectory + subDir)

        with Firefox(options=opts) as browser:
            # interact with duck duck
            browser.get(
                "http://"
                + depragIP
                + "/cgi-bin/cgiread?site=-&request=fvalues&args=&mode=-1-"
            )

            downloads = browser.find_elements(By.CLASS_NAME, "download")
            downloads[len(downloads) - 2].click()

        info = open(downloadDirectory + subDir + "/actual.csv", "r")
        _info = info.read()
        info.close()
        _info = _info[20:]
        c = _info.find(",")
        _info = _info[:c]

        _prgrmNumber = int(_info)
        print(_prgrmNumber)

        # get programm info
        _str = (
            "http://"
            + depragIP
            + "/cgi-bin/cgiread?site=getprg&program="
            + str(_prgrmNumber)
        )
        r = requester.get(_str)
        prgmInfo = r.text
        print(prgmInfo)  # for debuging can be disabled

        msg = screwing()
        msg.prgrmInfo = prgmInfo
        os.rename(
            downloadDirectory + "/" + _kurve_name, downloadDirectory + "/data.txt"
        )

        with open(downloadDirectory + "/data.txt", "r") as _csv:
            _csv_str = _csv.read()
        c = _csv_str.find("Temperatur")
        c = c + 15
        _csv_str = _csv_str[c:]
        msg.csv = _csv_str
        publer.publish(msg)

        filepath = backupDownload + "/" + str(time.time()) + ".json"

        with open(filepath, "w") as _f:
            jsonInfo = prgmInfo.replace("\n", ',\n"')
            jsonInfo = jsonInfo.replace("=", '":')
            jsonInfo = '"' + jsonInfo
            c = find_nth(jsonInfo, ":", 2) + 1
            jsonInfo = jsonInfo[:c] + '"' + jsonInfo[c:]
            c = find_nth(jsonInfo, ",", 2)
            jsonInfo = jsonInfo[:c] + '"' + jsonInfo[c:]
            _f.write("{ \n")
            _f.write(jsonInfo)

            zeit = "["
            Dmess1 = "["
            Wmess1 = "["
            Dmotor = "["
            Wmotor = "["
            DZ = "["
            Schritt = "["
            Strom = "["
            Temp = "["
            lines = _csv_str.count("\n") + 1  # anzahl Zeilen
            for i in range(0, lines):
                # Zeit
                c = _csv_str.find(",")
                zeit = zeit + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # Mess1 Dreh
                c = _csv_str.find(",")
                Dmess1 = Dmess1 + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # W mess1
                c = _csv_str.find(",")
                Wmess1 = Wmess1 + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # D motor
                c = _csv_str.find(",")
                Dmotor = Dmotor + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # W Motor
                c = _csv_str.find(",")
                Wmotor = Wmotor + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # Drehzahl
                c = _csv_str.find(",")
                DZ = DZ + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # Schritt
                c = _csv_str.find(",")
                Schritt = Schritt + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # Strom
                c = _csv_str.find(",")
                Strom = Strom + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]
                # temp
                c = _csv_str.find("\n")
                Temp = Temp + _csv_str[:c] + ","
                _csv_str = _csv_str[c + 1 :]

            zeit = zeit.replace("\n", "")

            zeit = zeit[: (len(zeit) - 2)] + "],"
            Dmess1 = Dmess1[: (len(zeit) - 2)] + "],"
            Wmess1 = Wmess1[: (len(zeit) - 2)] + "],"
            Dmotor = Dmotor[: (len(zeit) - 2)] + "],"
            Wmotor = Wmotor[: (len(zeit) - 2)] + "],"
            DZ = DZ[: (len(zeit) - 2)] + "],"
            Schritt = Schritt[: (len(zeit) - 2)] + "],"
            Strom = Strom[: (len(zeit) - 2)] + "],"
            Temp = Temp[: (len(zeit) - 2)] + "]"

            _f.write('csv" : { \n')
            t = '   "Zeit(0.001 ms)":' + zeit
            t.replace("\n", "")
            _f.write(t + "\n")
            _f.write('  "Drehmoment Messsystem 1 (Nm)":' + Dmess1 + "\n")
            _f.write('  "Winkel Messsystem 1":' + Wmess1 + "\n")
            _f.write('  "Drehmoment Motor (Nm)":' + Dmotor + "\n")
            _f.write('  "Winkel Motor":' + Wmotor + "\n")
            _f.write('  "Drehzahl (U/min)":' + DZ + "\n")
            _f.write('  "Schritt":' + Schritt + "\n")
            _f.write('  "Stromstaerke (A)":' + Strom + "\n")
            _f.write('  "Temperatur (C)":' + Temp + "\n")

            _f.write("}\n")
            _f.write("  }")

        shutil.rmtree(downloadDirectory, ignore_errors=True)

    downloadDirectory = backupDownload
    return []


publer = 0


def downloader():
    global publer
    rospy.init_node("deprag_downloader", anonymous=True)
    service_download = rospy.Service(
        "deprag_download", download_request, callback_download
    )
    publer = rospy.Publisher("deprag/curves", screwing, queue_size=10)

    rospy.spin()


if __name__ == "__main__":  # apparetly good practise in python
    try:
        downloader()
    except rospy.ROSInterruptException:
        pass
